### PR TITLE
Update froidure-pin-hpp for v1/3

### PR DESCRIPTION
**Pull request by james-d-mitchell**
This PR requires the corresponding one in libsemigroups to be merged first, and also for some of the tests to be adjusted.

### CODE REVIEW

File: libsemigroups_pybind11/detail/decorators.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                    parameters=(
+                        inspect.Parameter(
+                            "dummy", inspect.Parameter.POSITIONAL_ONLY
+                        ),
+                    )
+                    + tuple(val for val in sig.parameters.values())
+                )
+        wrapper.__signature__ = sig
+    except ValueError:
+        pass
```
Comment (james-d-mitchell) [inline libsemigroups_pybind11/detail/decorators.py:103]: @Joseph-Edwards the ridiculous fix

**Review by Joseph-Edwards — CHANGES_REQUESTED** [review]
All looking good to me. Modulo the suggestions outlined below, I'll be happy to merge. Thanks @james-d-mitchell!

File: src/froidure-pin.cpp [hunk 2]
```diff
@@ -0,0 +1,763 @@
...
+                [](FroidurePin_ const& self) { return FroidurePin_(self); });
+      thing.def(
+          "copy",
+          [](FroidurePin_ const& self) { return FroidurePin_(self); },
+          R"pbdoc(
+Copy a :any:`FroidurePinPBR` object.
+
+:returns: A copy.
+:rtype: FroidurePinPBR
+)pbdoc");
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply src/froidure-pin.cpp:119]: Yup, as discussed.

File: libsemigroups_pybind11/froidure_pin.py [hunk 3]
```diff
@@ -6,131 +6,384 @@
...
+    def generator(self: Self, i: int) -> Element:
+        return self._cxx_obj.generator(i)
+
+    @_returns_element
+    def sorted_at(self: Self, i: int) -> Element:
+        return self._cxx_obj.sorted_at(i)
+
+
+########################################################################
+# Helpers -- from froidure-pin.cpp
```
[the remark this answers is not in the document]
Comment (james-d-mitchell) [reply libsemigroups_pybind11/froidure_pin.py:181]: I think I've resolved this, some of the signature should say `FroidurePinBase` and I just got it wrong when I was writing them in this file. Hopefully caught them all!

**Comment by james-d-mitchell** [comment]
@Joseph-Edwards you happy for me to merge this?

**Comment by Joseph-Edwards** [comment]
> @Joseph-Edwards you happy for me to merge this?

Yeah looks good to me. Merge away! Thanks @james-d-mitchell!

**Pull request merged**